core/trino-main/src/main/java/io/trino/operator/join/unspilled/PartitionedLookupSource.java

```diff
@@ -144,6 +144,58 @@ public long getJoinPosition(int position, Page hashChannelsPage, Page allChannel
         return encodePartitionedJoinPosition(partition, toIntExact(joinPosition));
     }
 
+    @Override
+    public void getJoinPosition(int[] positions, Page hashChannelsPage, Page allChannelsPage, long[] rawHashes, long[] result)
+    {
+        int positionCount = positions.length;
+        int partitionCount = partitionGenerator.getPartitionCount();
+
+        int[] partitions = new int[positionCount];
+        int[] partitionPositionsCount = new int[partitionCount];
+
+        // Get the partitions for every position and calculate the size of every partition
+        for (int i = 0; i < positionCount; i++) {
+            int partition = partitionGenerator.getPartition(rawHashes[positions[i]]);
+            partitions[i] = partition;
+            partitionPositionsCount[partition]++;
+        }
+
+        int[][] positionsPerPartition = new int[partitionCount][];
```

Comment:
nit: you could just have one array like:
```
[0 .. positionCount-1]
```
and store positions in slices
```
[0.. partitionPositionsCount[0], partitionPositionsCount[0]+1, .. partitionPositionsCount[1], ...]
```

Reply:
Tried it, did not help.